### PR TITLE
Rework inherent data client side

### PR DESCRIPTION
**Pull request by bkchr**
This pr changes the inherent data system to be more flexible, especially the client side. Before this pr we only supported calling `create_inherent_data` and required that the inherent data was created without any context. We also shared the inherent data providers between block production and block import, while some inherents don't require any inherent data while being checked (at import).

So, the pr does the following high level changes:

- Split `create_inherent_data` into `create_inherent_data_providers` and `create_inherent_data`. While the former is async and thus, can use `await`.
- By default always pass the `parent_hash` to `create_inherent_data_providers` and add support for passing other arguments as well, while the caller can decide on which extra arguments to pass. This is especially useful for Cumulus that will for example also pass the `relay_chain_hash` a block will be build on.
- Move `ProvideInherent` to `frame-support`, because this is a FRAME specific trait and nothing generic.
- Reworked the inherent data providers to make the usage easier.
- Block building and block import is now separated, making it easier to decide which inherent data is required where.

polkadot companion: https://github.com/paritytech/polkadot/pull/2845

### CODE REVIEW

**Review by andresilva — COMMENTED** [review]
the changes lgtm. what was a bit weird to me is that now the `InherentDataProvider` is kind of a one-off thing, you just create it pre-filled with the data you want it to provide for a given block. My initial reaction when I read the PR description was: why wouldn't the parent hash be passed to `InherentDataProvider` instead of creating this new `CreateInherentDataProviders` trait? Centralizing the setup of inherent data providers is good though, instead of having these registrations sprinkled through the codebase.

**Comment by bkchr** [comment]
> what was a bit weird to me is that now the `InherentDataProvider` is kind of a one-off thing, you just create it pre-filled with the data you want it to provide for a given block. My initial reaction when I read the PR description was: why wouldn't the parent hash be passed to `InherentDataProvider` instead of creating this new `CreateInherentDataProviders` trait? 

I wanted to remove this "create InherentData and extract data later in the client again from the inherent data". I wanted to bake this into the type system to make it "more" obvious on what kind of inherent data providers are required by the "consensus" for example.

File: client/consensus/babe/src/tests.rs [hunk 1]
```diff
@@ -418,8 +437,8 @@ fn run_one_test(
 			// run each future until we get one of our own blocks with number higher than 5
 			// that was produced locally.
 			client.import_notification_stream()
-				.take_while(move |n| future::ready(n.header.number() < &5 || {
-					if n.origin == BlockOrigin::Own {
+				.take_while(move |n| future::ready(dbg!(n.header.number()) < &5 || {
+					if dbg!(n.origin) == BlockOrigin::Own {
```
Comment (rphmeier) [inline client/consensus/babe/src/tests.rs:441]: some stray `dbg!`

File: primitives/inherents/src/lib.rs [hunk 2]
```diff
@@ -15,21 +15,149 @@
...
+//!
+//! async fn cool_consensus_block_production(
+//! 	// The second parameter to the trait are parameters that depend on what the caller
+//! 	// can provide on extra data.
+//! 	_: impl sp_inherents::CreateInherentDataProviders<Block, ()>,
+//! ) {
+//! 	// do cool stuff
+//! }
+//!
+//! async fn cool_consensus_block_import(
```
Comment (rphmeier) [inline primitives/inherents/src/lib.rs:129]: This is supposed to model some block import/verification task?

**Review by rphmeier — APPROVED** [review]

File: primitives/inherents/src/lib.rs (hunk 2 above)
Comment (bkchr) [reply primitives/inherents/src/lib.rs:129]: Yeah some shitty import/verification task :D

File: primitives/consensus/common/src/import_queue/basic_queue.rs [hunk 3]
```diff
@@ -9,7 +9,7 @@
 //
 // 	http://www.apache.org/licenses/LICENSE-2.0
 //
-// Unless required by applicable law or agreed to in writing, software
+
```
Comment (andresilva) [inline primitives/consensus/common/src/import_queue/basic_queue.rs:12]:
```suggestion
// Unless required by applicable law or agreed to in writing, software
```

**Review by andresilva — APPROVED** [review]

**Comment by bkchr** [comment]
bot merge

**Pull request merged**